Generic type declaration should be mandatory in structs with embedded generic structs

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 18.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-2400 CPU @ 3.10GHz
CC version: cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

getwd: /Git/vlab
vmodules: /home/alejandro/.vmodules
vroot: /opt/v
vexe: /opt/v/v
vexe mtime: 2021-05-02 18:53:11
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 273655e.fb7ddcd

Git version: git version 2.17.1
Git vroot status: weekly.2021.17-60-gfb7ddcd4
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg example4.v`
```v
struct GenericChannelStruct<T> {
	ch chan T
}

struct MyGenericChannelStruct {
	GenericChannelStruct<T>
	msg string
}

struct Simple {
	msg string
}

fn main() {
	d := new_channel_struct<Simple>()
}

pub fn new_channel_struct<T>() GenericChannelStruct<T> {
	d := GenericChannelStruct{
		ch: chan T{}
	}

	return d
}

```

**What did you expect to see?**

I think a compiler error, as "MyGenericChannelStruct" includes "GenericChannelStruct" as an embedded struct, but "GenericChannelStruct" is a generic type, so "MyGenericChannelStruct" should include the generic type.

**What did you see instead?**
```
example4.v:15:2: warning: unused variable: `d`
   13 | 
   14 | fn main() {
   15 |     d := new_channel_struct<Simple>()
      |     ^
   16 | }
   17 |
==================
/tmp/v/example4.15379964027069431146.tmp.c:1078: error: ';' expected (got "main__GenericChannelStruct_T_T")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -523,8 +523,12 @@ pub fn (mut c Checker) struct_decl(mut decl ast.StructDecl) {
 		c.check_valid_pascal_case(decl.name, 'struct name', decl.pos)
 	}
 	mut struct_sym := c.table.find_type(decl.name) or { ast.TypeSymbol{} }
+	mut has_generic_types := false
 	if mut struct_sym.info is ast.Struct {
 		for embed in decl.embeds {
+			if embed.typ.has_flag(.generic) {
+				has_generic_types = true
+			}
 			embed_sym := c.table.get_type_symbol(embed.typ)
 			if embed_sym.kind != .struct_ {
 				c.error('`$embed_sym.name` is not a struct', embed.pos)
@@ -542,6 +546,9 @@ pub fn (mut c Checker) struct_decl(mut decl ast.StructDecl) {
 		}
 		for i, field in decl.fields {
 			c.ensure_type_exists(field.typ, field.type_pos) or { return }
+			if field.typ.has_flag(.generic) {
+				has_generic_types = true
+			}
 			if decl.language == .v {
 				c.check_valid_snake_case(field.name, 'field name', field.pos)
 			}
@@ -592,6 +599,10 @@ pub fn (mut c Checker) struct_decl(mut decl ast.StructDecl) {
 					}
 				}
 			}
+		}
+		if decl.generic_types.len == 0 && has_generic_types {
+			c.error('generic struct declaration must specify the generic type names, e.g. Foo<T>',
+				decl.pos)
 		}
 	}
 }

--- a/vlib/v/checker/tests/generics_struct_declaration_err.out
+++ b/vlib/v/checker/tests/generics_struct_declaration_err.out
@@ -5,10 +5,3 @@ vlib/v/checker/tests/generics_struct_declaration_err.vv:5:1: error: generic stru
       | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     6 |     GenericChannelStruct<T>
     7 |     msg string
-vlib/v/checker/tests/generics_struct_declaration_err.vv:15:2: error: unused variable: `d`
-   13 |
-   14 | fn main() {
-   15 |     d := new_channel_struct<Simple>()
-      |     ^
-   16 | }
-   17 |

--- a/vlib/v/checker/tests/generics_struct_declaration_err.out
+++ b/vlib/v/checker/tests/generics_struct_declaration_err.out
@@ -1,0 +1,14 @@
+vlib/v/checker/tests/generics_struct_declaration_err.vv:5:1: error: generic struct declaration must specify the generic type names, e.g. Foo<T>
+    3 | }
+    4 |
+    5 | struct MyGenericChannelStruct {
+      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    6 |     GenericChannelStruct<T>
+    7 |     msg string
+vlib/v/checker/tests/generics_struct_declaration_err.vv:15:2: error: unused variable: `d`
+   13 |
+   14 | fn main() {
+   15 |     d := new_channel_struct<Simple>()
+      |     ^
+   16 | }
+   17 |

--- a/vlib/v/checker/tests/generics_struct_declaration_err.vv
+++ b/vlib/v/checker/tests/generics_struct_declaration_err.vv
@@ -12,7 +12,7 @@ struct Simple {
 }
 
 fn main() {
-	d := new_channel_struct<Simple>()
+	new_channel_struct<Simple>()
 }
 
 pub fn new_channel_struct<T>() GenericChannelStruct<T> {

--- a/vlib/v/checker/tests/generics_struct_declaration_err.vv
+++ b/vlib/v/checker/tests/generics_struct_declaration_err.vv
@@ -1,0 +1,24 @@
+struct GenericChannelStruct<T> {
+	ch chan T
+}
+
+struct MyGenericChannelStruct {
+	GenericChannelStruct<T>
+	msg string
+}
+
+struct Simple {
+	msg string
+}
+
+fn main() {
+	d := new_channel_struct<Simple>()
+}
+
+pub fn new_channel_struct<T>() GenericChannelStruct<T> {
+	d := GenericChannelStruct{
+		ch: chan T{}
+	}
+
+	return d
+}

